elide-datastore/elide-datastore-aggregation/src/test/java/com/yahoo/elide/datastores/aggregation/queryengines/sql/expression/JoinExpressionExtractorTest.java

```diff
@@ -0,0 +1,541 @@
+        dim4Arg.put("joinArg4", Argument.builder().name("joinArg4").value("foo4").build());
+        SQLDimensionProjection dim4a = (SQLDimensionProjection) table.getDimensionProjection("dim4a", "dim4a", dim4Arg);
+
+        Map<String, Argument> dim5Arg = new HashMap<>();
+        dim5Arg.put("exprArg", Argument.builder().name("exprArg").value("same").build());
+        dim5Arg.put("joinArg4", Argument.builder().name("joinArg4").value("foo5").build());
+        SQLDimensionProjection dim5a = (SQLDimensionProjection) table.getDimensionProjection("dim5a", "dim5a", dim5Arg);
+
+        Query query = Query.builder()
+                        .source(table)
+                        .dimensionProjection(dim4a)
+                        .dimensionProjection(dim5a)
+                        .arguments(emptyMap())
+                        .build();
+
+        String generatedSql = engine.explain(query).get(0);
+
+        String expectedSQL =
+                          "SELECT " + NL
+                        + "  DISTINCT `MainTable_join2a_XXX`.`dim1` AS `dim4a`," + NL
+                        + "           `MainTable_join2a_XXX`.`dim1` AS `dim5a` " + NL
+                        + "FROM " + NL
+                        + "  `main_table` AS `MainTable` " + NL
+                        + "LEFT OUTER JOIN " + NL
+                        + "  `join_table` AS `MainTable_join2a_XXX` " + NL
+                        + "ON " + NL
+                        + "  value = 'same' AND `MainTable`.`dim6` - 'bar' = `MainTable_join2a_XXX`.`dim4` - 'fixedFoo' ";
+
+        assertEquals(formatExpected(expectedSQL), formatGenerated(generatedSql));
+    }
+
+    private String formatExpected(String expectedSQL) {
+        // Remove spaces at the start of each line with in string
+        expectedSQL = expectedSQL.replaceAll("(?m)^\\s*", "");
+        expectedSQL = expectedSQL.replace(NL, "");
+        return REPEATEDSPACE_PATTERN.matcher(expectedSQL).replaceAll(" ");
+    }
+
+    private String formatGenerated(String generatedSql) {
+        generatedSql = generatedSql.replace(ALIAS_PREFIX, "");
+        generatedSql = REPEATEDSPACE_PATTERN.matcher(generatedSql).replaceAll(" ");
+        return replaceDynamicAliases(generatedSql);
+    }
+}
+
+
+@EqualsAndHashCode
+@ToString
+@Data
+@FromTable(name = "main_table")
+@Include(name = "mainTable")
+class MainTable {
+
+    @Id
+    private String id;
+
+    @DimensionFormula("{{join.dim1}}")
+    private String dim1;
+
+    @DimensionFormula("{{join.dim2}}")
+    private String dim2;
+
+    @DimensionFormula("{{join.joinjoin.dim2}}")
+    private String dim3;
+
+    @Join("{{$id}} = {{join.$id}}")
+    private JoinTable join;
+
+    @DimensionFormula(value = "{{join2.dim1}}",
+                      arguments = { @ArgumentDefinition(name = "exprArg", type = ValueType.TEXT),
+                                    @ArgumentDefinition(name = "joinArg4", type = ValueType.TEXT)})
```

Comment:
The 4 in joinArg4 is confusing.  Do the numbers mean anything?

Reply:
renamed to joinArg